File: SERVER/go_core/ingestion/site_partitioner.go
package ingestion

import (
	"fmt"
	"regexp"
	"strings"
)

// SitePartitionRoute defines the site-partitioned NATS subject structure (allsite.md)
type SitePartitionRoute struct {
	SiteID     string `json:"site_id"`
	Subject    string `json:"subject"`
	Category   string `json:"category"`   // telemetry, incident, approval
	Severity   string `json:"severity"`   // critical, warning, normal
	OrderKey   string `json:"order_key"`  // per-site ordering key
}

var nonAlphaNumRegex = regexp.MustCompile(`[^a-zA-Z0-9_-]`)

// NormalizeSiteID cleanses and standardizes raw site identifiers into NATS-compliant subject tokens
func NormalizeSiteID(rawSite string) string {
	rawSite = strings.TrimSpace(rawSite)
	if rawSite == "" {
		return "default-site"
	}
	
	// Convert spaces to hyphens and remove invalid characters
	clean := strings.ReplaceAll(strings.ToLower(rawSite), " ", "-")
	clean = nonAlphaNumRegex.ReplaceAllString(clean, "")
	if clean == "" {
		return "default-site"
	}
	return clean
}

// GetPartitionedSubject resolves the exact NATS JetStream subject name per site & severity
func GetPartitionedSubject(category string, siteID string, severity string) string {
	cleanSite := NormalizeSiteID(siteID)
	cleanSev := strings.ToLower(strings.TrimSpace(severity))

	switch strings.ToLower(category) {
	case "telemetry":
		if cleanSev == "critical" || cleanSev == "error" || cleanSev == "fatal" {
			return fmt.Sprintf("telemetry.site.%s.critical", cleanSite)
		} else if cleanSev == "warning" || cleanSev == "warn" {
			return fmt.Sprintf("telemetry.site.%s.warning", cleanSite)
		} else {
			return fmt.Sprintf("telemetry.site.%s.normal", cleanSite)
		}
	case "incident":
		if cleanSev == "update" {
			return fmt.Sprintf("incident.site.%s.update", cleanSite)
		}
		return fmt.Sprintf("incident.site.%s.create", cleanSite)
	case "approval":
		return fmt.Sprintf("approval.site.%s", cleanSite)
	default:
		return fmt.Sprintf("telemetry.site.%s.normal", cleanSite)
	}
}

// GetPartitionWildcards returns wildcard subjects for NATS consumers and metrics monitoring
func GetPartitionWildcards() []map[string]string {
	return []map[string]string{
		{"subject": "telemetry.site.*.critical", "role": "Site Critical Ingest Stream", "category": "Telemetry"},
		{"subject": "telemetry.site.*.warning", "role": "Site Warning Ingest Stream", "category": "Telemetry"},
		{"subject": "telemetry.site.*.normal", "role": "Site Normal Ingest Stream", "category": "Telemetry"},
		{"subject": "incident.site.*.create", "role": "Site Incident Creation Queue", "category": "Incident"},
		{"subject": "incident.site.*.update", "role": "Site Incident Update Channel", "category": "Incident"},
		{"subject": "approval.site.*", "role": "Site HITL Approval Channel", "category": "Approval"},
	}
}

File: SERVER/python_ai_core/telemetry/site_partitioner.py
"""
Python Site Partitioner Helper (allsite.md)
Provides NATS JetStream subject routing per site_id:
- telemetry.site.<site_id>.critical
- telemetry.site.<site_id>.warning
- telemetry.site.<site_id>.normal
- incident.site.<site_id>.create
- approval.site.<site_id>
"""

import re
from typing import Dict, Any, List

def normalize_site_id(raw_site: str) -> str:
    """Normalize raw site string to NATS-safe subject token."""
    if not raw_site or not raw_site.strip():
        return "default-site"
    clean = raw_site.strip().lower().replace(" ", "-")
    clean = re.sub(r'[^a-z0-9_-]', '', clean)
    return clean if clean else "default-site"

def get_partitioned_subject(category: str, site_id: str, severity: str = "normal") -> str:
    """Construct site-partitioned NATS subject string."""
    clean_site = normalize_site_id(site_id)
    clean_sev = severity.strip().lower()
    cat = category.strip().lower()

    if cat == "telemetry":
        if clean_sev in ["critical", "error", "fatal"]:
            return f"telemetry.site.{clean_site}.critical"
        elif clean_sev in ["warning", "warn"]:
            return f"telemetry.site.{clean_site}.warning"
        else:
            return f"telemetry.site.{clean_site}.normal"
    elif cat == "incident":
        if clean_sev == "update":
            return f"incident.site.{clean_site}.update"
        return f"incident.site.{clean_site}.create"
    elif cat == "approval":
        return f"approval.site.{clean_site}"
    else:
        return f"telemetry.site.{clean_site}.normal"

def get_all_site_wildcards() -> List[Dict[str, str]]:
    """Return wildcard subjects for telemetry streams."""
    return [
        {"subject": "telemetry.site.*.critical", "role": "Site Critical Ingest Stream"},
        {"subject": "telemetry.site.*.warning", "role": "Site Warning Ingest Stream"},
        {"subject": "telemetry.site.*.normal", "role": "Site Normal Ingest Stream"},
        {"subject": "incident.site.*.create", "role": "Site Incident Creation Queue"},
        {"subject": "incident.site.*.update", "role": "Site Incident Update Channel"},
        {"subject": "approval.site.*", "role": "Site HITL Approval Channel"}
    ]

def get_nats_reconnection_opts() -> Dict[str, Any]:
    """Return resilient reconnection options for multi-site WAN NATS connections."""
    return {
        "max_reconnect_attempts": 10,
        "reconnect_time_wait": 2.0,
        "ping_interval": 15,
        "max_outstanding_pings": 3,
        "tls_handshake_timeout": 10.0
    }

if __name__ == "__main__":
    print("Default Site:", normalize_site_id("  Kantor Cabang Jakarta #01 "))
    print("Subject Critical:", get_partitioned_subject("telemetry", "Kantor Cabang Jakarta", "CRITICAL"))
    print("Subject Incident:", get_partitioned_subject("incident", "Kantor Pusat", "NEW"))
    print("NATS Reconnect Opts:", get_nats_reconnection_opts())
